docker/src/main/java/brooklyn/location/docker/DockerContainerLocation.java

```diff
@@ -168,20 +171,57 @@ public HostAndPort getSocketEndpointFor(Cidr accessor, int privatePort) {
-            parseDockerCallback(commandString);
+        if(getOwner().config().get(DockerContainer.DOCKER_USE_EXEC)) {
+            Map<String,?> nonPortProps = Maps.filterKeys(props, Predicates.not(Predicates.containsPattern("port")));
+            SshMachineLocation host = getOwner().getDockerHost().getDynamicLocation().getMachine();
+            return host.execCommands(nonPortProps, summaryForLogging, getExecScript(commands, env));
+        } else {
+            Iterable<String> filtered = Iterables.filter(commands, DockerCallbacks.FILTER);
+            for (String commandString : filtered) {
+                parseDockerCallback(commandString);
+            }
+            return super.execScript(props, summaryForLogging, commands, env);
         }
-        return super.execScript(props, summaryForLogging, commands, env);
     }
 
     @Override
     public int execCommands(Map<String,?> props, String summaryForLogging, List<String> commands, Map<String,?> env) {
-        Iterable<String> filtered = Iterables.filter(commands, DockerCallbacks.FILTER);
-        for (String commandString : filtered) {
-            parseDockerCallback(commandString);
+        if(getOwner().config().get(DockerContainer.DOCKER_USE_EXEC)) {
+
+            SshMachineLocation host = getOwner().getDockerHost().getDynamicLocation().getMachine();
+            Map<String,?> nonPortProps = Maps.filterKeys(props, Predicates.not(Predicates.containsPattern("port")));
+            return host.execCommands(nonPortProps, summaryForLogging, getExecCommands(commands, env));
+        } else {
+            Iterable<String> filtered = Iterables.filter(commands, DockerCallbacks.FILTER);
+            for (String commandString : filtered) {
+                parseDockerCallback(commandString);
+            }
+            return super.execCommands(props, summaryForLogging, commands, env);
         }
-        return super.execCommands(props, summaryForLogging, commands, env);
+    }
+
+    private List<String> getExecScript(List<String> commands, Map<String,?> env) {
+        String prefix = "docker exec " + dockerContainer.getContainerId() + " '";
+        return Collections.singletonList(prefix + Joiner.on(';').join(
+                Iterables.concat(getEnvVarCommands(env), commands)) + "'");
+    }
+
+    private List<String> getExecCommands(List<String> commands, Map<String,?> env) {
+        List<String> result = new LinkedList<String>();
+        result.addAll(getEnvVarCommands(env));
+        String prefix = "docker exec " + dockerContainer.getContainerId() + " ";
+        for(String command: commands) {
+            result.add(prefix + command);
+        }
+        return result;
+    }
+
+    private List<String> getEnvVarCommands(Map<String,?> env) {
+        List<String> result = new LinkedList<String>();
+        for(Map.Entry<String, ?> envVar : env.entrySet()) {
+            result.add("export " + envVar.getKey() + "=" + envVar.getValue());
```

Comment:
See `ShellAbstractTool` for an example of escaping environment variables using `BashStringEscapes`